java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java

```diff
@@ -0,0 +1,515 @@
+package org.apache.arrow.memory;
+
+import io.netty.buffer.ArrowBuf;
+import io.netty.buffer.UnsafeDirectLittleEndian;
+import org.apache.arrow.memory.util.HistoricalLog;
+import org.apache.arrow.util.Preconditions;
+
+import java.util.IdentityHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static org.apache.arrow.memory.BaseAllocator.indent;
+
+/**
+ * The reference manager that binds an {@link AllocationManager} to
+ * {@link BufferAllocator} and a set of {@link ArrowBuf}. The set of
+ * ArrowBufs managed by this reference manager share a common
+ * fate (same reference count)
+ *
+ * As with AllocationManager, the only reason this is public is due to ArrowBuf
+ * being in io.netty.buffer package.
+ */
+public class BufferLedger implements ValueWithKeyIncluded<BaseAllocator>, ReferenceManager  {
+
+  private final IdentityHashMap<ArrowBuf, Integer> buffers = BaseAllocator.DEBUG ? new IdentityHashMap<>() : null;
+  private static final AtomicLong LEDGER_ID_GENERATOR = new AtomicLong(0);
+  // unique ID assigned to each ledger
+  private final long ledgerId = LEDGER_ID_GENERATOR.incrementAndGet();
+  private final AtomicInteger bufRefCnt = new AtomicInteger(0); // start at zero so we can
+  // manage request for retain
+  // correctly
+  private final long lCreationTime = System.nanoTime();
+  private final BaseAllocator allocator;
+  private final AllocationManager allocationManager;
+  private final HistoricalLog historicalLog =
+    BaseAllocator.DEBUG ? new HistoricalLog(BaseAllocator.DEBUG_LOG_LENGTH, "BufferLedger[%d]", 1) : null;
+  private volatile long lDestructionTime = 0;
+
+  BufferLedger(final BaseAllocator allocator, final AllocationManager allocationManager) {
+    this.allocator = allocator;
+    this.allocationManager = allocationManager;
+  }
+
+  boolean isOwningLedger() {
+    return this == allocationManager.getOwningLedger();
+  }
+
+  public BaseAllocator getKey() {
+    return allocator;
+  }
+
+  /**
+   * Get the buffer allocator associated with this reference manager
+   * @return buffer allocator
+   */
+  @Override
+  public BufferAllocator getAllocator() {
+    return allocator;
+  }
+
+  /**
+   * Get this ledger's reference count
+   * @return reference count
+   */
+  @Override
+  public int getRefCount() {
+    return bufRefCnt.get();
+  }
+
+  /**
+   * Increment the ledger's reference count for the associated
+   * underlying memory chunk. All ArrowBufs managed by this ledger
+   * will share the ref count.
+   */
+  void increment() {
+    bufRefCnt.incrementAndGet();
```

Comment:
could it be helpful to return the new count?

Reply:
I guess why do you need this since `retain()` is below?

Reply:
These interfaces are borrowed from how they were earlier in ArrowBuf. I didn't want to change or remove this reference count related APIs

Reply:
Ok, then I agree not to change it. I thought it was a new api.